br/pkg/lightning/errormanager/errormanager.go

```diff
@@ -288,38 +301,89 @@ func (em *ErrorManager) RecordIndexConflictError(
 				rawHandles[i],
 				rawRows[i],
 			)
-			if err != nil {
-				return err
-			}
 		}
-		return nil
+		_, err := txn.ExecContext(c, buf.String())
+		return err
 	})
 }
 
-// GetConflictKeys obtains all (distinct) conflicting rows (handle and their
-// values) from the current error report.
-func (em *ErrorManager) GetConflictKeys(ctx context.Context, tableName string, prevRowID int64, limit int) (handleRows [][2][]byte, lastRowID int64, err error) {
+// ResolveAllConflictKeys query all conflicting rows (handle and their
+// values) from the current error report and resolve them concurrently.
+func (em *ErrorManager) ResolveAllConflictKeys(
+	ctx context.Context,
+	tableName string,
+	pool *utils.WorkerPool,
+	fn func(ctx context.Context, handleRows [][2][]byte) error,
+) error {
 	if em.db == nil {
-		return nil, 0, nil
-	}
-	rows, err := em.db.QueryContext(
-		ctx,
-		fmt.Sprintf(selectConflictKeys, em.schemaEscaped),
-		tableName,
-		prevRowID,
-		limit,
-	)
-	if err != nil {
-		return nil, 0, errors.Trace(err)
+		return nil
 	}
-	defer rows.Close()
 
-	for rows.Next() {
-		var handleRow [2][]byte
-		if err := rows.Scan(&lastRowID, &handleRow[0], &handleRow[1]); err != nil {
-			return nil, 0, errors.Trace(err)
-		}
-		handleRows = append(handleRows, handleRow)
+	const rowLimit = 1000
+	taskCh := make(chan [2]int64)
+	taskWg := &sync.WaitGroup{}
+	g, gCtx := errgroup.WithContext(ctx)
+
+	go func() {
+		//nolint:staticcheck
+		taskWg.Add(1)
+		taskCh <- [2]int64{0, math.MaxInt64}
+		taskWg.Wait()
+		close(taskCh)
+	}()
+
+	for t := range taskCh {
+		start, end := t[0], t[1]
+		pool.ApplyOnErrorGroup(g, func() error {
+			defer taskWg.Done()
+
+			var handleRows [][2][]byte
+			for start < end {
+				rows, err := em.db.QueryContext(
+					gCtx, fmt.Sprintf(selectConflictKeys, em.schemaEscaped),
+					tableName, start, end, rowLimit)
+				if err != nil {
+					return errors.Trace(err)
+				}
+				var lastRowID int64
+				for rows.Next() {
+					var handleRow [2][]byte
+					if err := rows.Scan(&lastRowID, &handleRow[0], &handleRow[1]); err != nil {
+						return errors.Trace(err)
+					}
+					handleRows = append(handleRows, handleRow)
+				}
+				if err := rows.Err(); err != nil {
+					return errors.Trace(err)
+				}
+				if err := rows.Close(); err != nil {
+					return errors.Trace(err)
+				}
+				if len(handleRows) == 0 {
+					break
+				}
+
+				// If the remaining tasks cannot be processed at once, split the task
+				// into two subtasks and send one of them to the other idle worker if possible.
+				start = lastRowID + 1
+				if end-start > rowLimit {
```

Comment:
Since the init end value is `math.MaxInt64`, this split by half strategy seems not very effective because the upper half won't match any record for many rounds. I think we should either use `select min(_tidb_rowid), max(_tidb_rowid)` or do region sample like dumpling to determine the chunk boundries.

Reply:
> the upper half won't match any record

Yes. But there are at most 64 rounds. It won't take too long at a round if it doesn't match any records.
I did a test. The test shows the rowid range shrinks very quckly.
```
start 4611686018427388404 end 9223372036854775807 cost 2.76702ms
start 2305843009213695202 end 4611686018427388404 cost 3.039887ms
start 1152921504606849101 end 2305843009213695202 cost 3.402173ms
start 576460752303426551 end 1152921504606849101 cost 2.67341ms
start 288230376151715776 end 576460752303426551 cost 2.323419ms
start 144115188075860888 end 288230376151715776 cost 2.351432ms
start 72057594037933944 end 144115188075860888 cost 2.078845ms
start 36028797018970972 end 72057594037933944 cost 2.300226ms
start 18014398509489986 end 36028797018970972 cost 2.856089ms
start 9007199254749993 end 18014398509489986 cost 2.932793ms
start 4503599627380497 end 9007199254749993 cost 3.60532ms
start 2251799813696249 end 4503599627380497 cost 2.458385ms
start 1125899906854625 end 2251799813696249 cost 2.39209ms
start 562949953434313 end 1125899906854625 cost 3.313871ms
start 281474976724657 end 562949953434313 cost 2.331175ms
start 140737488370329 end 281474976724657 cost 3.385059ms
start 70368744193665 end 140737488370329 cost 2.538585ms
start 35184372105833 end 70368744193665 cost 2.309589ms
start 17592186062417 end 35184372105833 cost 2.645189ms
start 8796093041209 end 17592186062417 cost 3.281738ms
start 4398046535865 end 8796093041209 cost 5.143813ms
start 2199023283693 end 4398046535865 cost 3.055257ms
start 1099511658107 end 2199023283693 cost 2.539213ms
start 549755845814 end 1099511658107 cost 3.101434ms
start 274877940167 end 549755845814 cost 2.494364ms
start 137438987844 end 274877940167 cost 1.994178ms
start 68719512182 end 137438987844 cost 2.626676ms
start 34359774851 end 68719512182 cost 3.69509ms
start 17179906686 end 34359774851 cost 2.520491ms
start 8589973103 end 17179906686 cost 2.939711ms
start 4295006812 end 8589973103 cost 3.160254ms
start 2147524166 end 4295006812 cost 4.009733ms
start 1073783343 end 2147524166 cost 2.513925ms
start 536913432 end 1073783343 cost 5.089464ms
start 268478976 end 536913432 cost 3.635151ms
start 134262248 end 268478976 cost 2.953334ms
start 67154384 end 134262248 cost 2.51085ms
start 33600952 end 67154384 cost 3.332525ms
start 16824736 end 33600952 cost 2.854554ms
start 12631432 end 16824736 cost 2.38818ms
start 10535280 end 12631432 cost 3.152013ms
start 9487704 end 10535280 cost 8.250281ms
```

Reply:
`select min(_tidb_rowid), max(_tidb_rowid)` is also useful optimization. But I'm not sure how it this query will take.

My original implementation is use region sample. But because we disable PD scheduler, the region sample didn't work well.

Reply:
A test shows querying min/max rowid costs 825.925012ms. It seems it doesn't worth doing this query in advance.